### ISSUE TITLE
Q3: Update issue history in Domo from issues in Github/Zenhub

### ISSUE THREAD

**Issue by BruceP99166** [openstax/domo#60]
@BruceP99166 commented on [Wed Oct 03 2018](https://github.com/openstax/organizational-agility/issues/32)

# Description
Update Domo issue_history with data from Github/Zenhub

# Acceptance
* Today's issue data from Github/Zenhub appended to ghzh_issue_history

# Solution
Python script to read github/zenhub and create ghzh_history datasets in Domo.  

# Tests
- Fields included: organization_name, repo_name, pipeline_name, issue_title, issue_number, child_issue_numbers, issue_labels, issue_releases, issue_url, open_closed, issue_estimate, issue_is_epic, issue_is_pr, repo_id, batch_id, batch_run
- batch_id contains the current max batch_id + 1
- batch_run contains the update datetime and is the same for all of today's rows
- All issues and PRs are included for each repo except marker issues with '---' in title

# Tasks
- [x] Explore zenhub api
- [x] Refactor burnup dataflow
- [x] Create ghzh_history_issue in Domo
- [x] Include all issues in the openstax and cnx organizations
- [x] Change fieldnames to lower case.
- [x] Include PRs in update.  
- [x] Include child_issue_number to store child issues of epics. 
- [ ] Fix error causing "Update failed" popup
- [ ] Document the code
- [x] Productionize

---

@BruceP99166 commented on [Tue Oct 09 2018](https://github.com/openstax/organizational-agility/issues/32#issuecomment-428231589)

Domo json connector

BoardName = not needed
ListName = zenhub issues pipeline/name
CardName = github issues title
Label Names = github issues labels
Card ShortUrl = github issues url
Card Closed = not used
Points = zenhub issues estimate/value
BoardId = not needed
ListId = not available
CardId = github issues node_id
Join = "y"
BatchId = not available
BatchLastRun = generated

---

@BruceP99166 commented on [Fri Oct 26 2018](https://github.com/openstax/organizational-agility/issues/32#issuecomment-433540787)

Zenhub api for board info does not return closed issues therefore cannot get estimates for them.
Can request issue by issue but this will exceed api rate limit.

Solution for closed issues is to look up in ghzh_issue_history the estimate when the issue was last open.

If an issue is opened and closed in one day, it may never have an estimate in ghzh_issue_history.  Then will set estimate to None.

---

@BruceP99166 commented on [Wed Jan 02 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-450995242)

GhZhToDomo 2019-01-02 15:36
Failed at 53:21 elapsed time with http 408
Request URL: https://api.zenhub.io/p1/reports/release/5bbe1c639c41fb34bb46ae07/issues?access_token=aea0232386bd74d064d4403c7f7dcf7c78d826763720efe074ea0231b9a445c3a434512da822e6b0

Rerunning with status retry specified for http session
retry = Retry(connect=3, status=3, status_forcelist=([408]), backoff_factor=0.5)

<blockquote></blockquote>

---

@BruceP99166 commented on [Tue Feb 05 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-458569044)

Successfully created ghzh_issue from GitHub/ZenHub issues but unable to use it as input to a dataflow in Domo.
(https://dojo.domo.com/t5/Beast-Mode-ETL-Dataflow/Unable-to-load-this-DataSet/td-p/11720 says this can be caused by 
- dataset name too long )
- This error is usually caused by one of the following:
    You may have one or more column header names that exceed the 64-character limit.
    You may have row level data in one or more columns that exceeds the 1024-character limit.

Try exporting GH issues to csv and creating ghzh_a1 ds from that to reproduce error.
- Processed records 0 - 1580.  No error.
- Processed records 1581-3000. No error.
- Processed records 3001-4176. No error.
- Processing records 4176-5000. No error.
- Processing records 5001-6000. No error.
- Processing records 6001-7000. No error.
- Processed records 7001-8553.  No error but output ds only has 1520 records

Will suggests that the problem is due to records with long issue names.  
The long record is in Work Management repo; test by running with just this repo.  Seems to work with issue names < 1024 but the test was not conclusive.

Now run GhZhToDomo with subsets of repos and use the created ghzh-issue as input to a dataflow.
- a15k-deployment to CC-LO-Maps: 473 issues.  Ran ok.
- cnx to cnx-varnish: 1307 issues.  Ran ok.
- CollegePhysicsApp to oer.exports: Running
- oer-simulation-study to research-survey-plans.  676 issues.  Ran ok.
- resque-status to work-management-reports.  1895 issues.  Ran ok.

Try running GhZhToDomo with all repos.  ghzh_issues created but gives same error message when used as dataflow input.
Emailed Will with the question.

Try running GhZhToDomo for subsets of the repos, then in Domo concatenating them into a single ghzh_issues for input to dataflow [GhZh Update Issue History].
- GhZhToDomo1: repos 0 to 50 > ghzh_issue1: 5256 issues; max fieldlen is IssueName=758.  Ran ok. Dataflow input fails.
- GhZhToDomo2: repos 51 to 100 > ghzh_issue2: 1298 issues; max fieldlen is IssueName=274.  Ran ok.  Dataflows input ok.
- GhZhToDomo3: repos 101 to 150 > ghzh_issue3: 265 issues processed but ghzh_issue3 has 264 records; max fieldlen is IssueName=363.  Ran ok. Dataflow input ok.
- GhZhToDomo4: repos 151 to 200 > ghzh_issue4: 1693 issues; max fieldlen: IssueName=229.  Ran ok.  Dataflow input ok.
- GhZhToDomo5: repos 200 to 250 > ghzh_issue5: 14 issues; max fieldlen: IssueName=93.  Ran ok. Dataflow input ok.

Look at repos 0-50  https://docs.google.com/spreadsheets/d/1TKgHQFrNcqdjScea4EG1zMZZRh-zKXwdWCMNdpkhAO8/edit#gid=0
- GhZhToDomo1 repos 0-10.  Export stopped at repo rhaptos.cnxmutiols issue 6.  Export ran ok.  Dataflow input ran ok.
- GhZhToDomo2 repos 11-25.  Export ran ok.  1158 issues.  Dataflow input ok.
- GhZhToDomo3 repos 26-50.  Export ran ok: 734 issues.  Dataflow input ok.

All the repo subsets have now successfully exported and been used as dataflow input.  
Will now try a complete update using subsets

190204
GhZhToDomo: repos 0 to 9999 > ghzh_issue 592275c1-a946-4a3f-a230-c05c41de74ac: issues processed=8553 output=8532
Appeared to run successfully; try as input in GhZh a1 dataflow.  Failed with same error.

---

@BruceP99166 commented on [Tue Jan 29 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-458588358)

ds ghzh_batchid does not seem to be updating, so workflow GhZh Update Issue History is not triggered.

Reinitialized ghzh_batchid; seems ok now.

---

@BruceP99166 commented on [Tue Feb 05 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-459779844)

Set up dataflows to update ghzh_issue_history from ghzh_issue1-4
Added command line parm to GhZhToDomo.py to allow specifying the dsid of the issue file to be created.
[GhZh Concatenate Issue] to concatenate GhZhToDomo1-4.  
Inputs
- GhZhToDomo1: repos 0 to 10 > ghzh_issue1 6cd67468-4d9e-41cf-a135-832226a3f012: issues=3367
- GhZhToDomo2: repos 11 to 100 > ghzh_issue2 197b88a0-f8b3-4d97-af8d-27231b5f7f3c: issues=3195
- GhZhToDomo3: repos 101 to 150 > ghzh_issue3 a9f405b3-2958-4274-a93c-7ba3d045f543: issues=266
- GhZhToDomo4: repos 151 to 300 > ghzh_issue4 888d6fb1-3b13-4626-b417-a6a86c1b1296: issues=1710

Output
- ghzh_issue bc2cb370-249c-4401-97b1-5471008a97ec

[GhZh Update Issue History] to update ghzh_issue_history.  Ran successfully.
Input
- ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59
- ghzh_batchid 4e88a075-3fd1-4e57-9771-91d94ba90b42
- ghzh_issue bc2cb370-249c-4401-97b1-5471008a97ec

Output
- ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59
- ghzh_batchid 4e88a075-3fd1-4e57-9771-91d94ba90b42

GhZhToDomo1-4 ran ok
[GhZh Concatenate Issue] ran ok
[GhZh Update Issue History] ran ok.

Rerun with 4 repo subsets
GhZhToDomo ran ok but did not output all issues processed for 1 & 2.
    GhZhToDomo1: repos 0 to 10 > ghzh_issue1 6cd67468-4d9e-41cf-a135-832226a3f012: issues processed=3367 output=3365
    GhZhToDomo2: repos 11 to 100 > ghzh_issue2 197b88a0-f8b3-4d97-af8d-27231b5f7f3c: issues processed=3198 output=3194
    GhZhToDomo3: repos 101 to 150 > ghzh_issue3 a9f405b3-2958-4274-a93c-7ba3d045f543: issues processed=268 output=268
    GhZhToDomo4: repos 151 to 300 > ghzh_issue4 888d6fb1-3b13-4626-b417-a6a86c1b1296: issues processed=1711 output=1711

[GhZh Concatenate Issue]
- ghzh_issue bc2cb370-249c-4401-97b1-5471008a97ec issues=8538

[GhZh Update Issue History]
Input
- ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59 issues=8702
- ghzh_batchid 4e88a075-3fd1-4e57-9771-91d94ba90b42 value=16
- ghzh_issue bc2cb370-249c-4401-97b1-5471008a97ec issues=8538

Output
- ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59 issues=17240
- ghzh_batchid 4e88a075-3fd1-4e57-9771-91d94ba90b42 value=17

Rerun with 3 repo subsets
GhZhToDomo ran ok but did not output all issues processed for 1 & 2.
    GhZhToDomo1: repos 0 to 10 > ghzh_issue1 6cd67468-4d9e-41cf-a135-832226a3f012: issues processed=3367 output=3366
    GhZhToDomo2: repos 11 to 100 > ghzh_issue2 197b88a0-f8b3-4d97-af8d-27231b5f7f3c: issues processed=3198 output=3198
    GhZhToDomo3: repos 101 to 300- > ghzh_issue3 a9f405b3-2958-4274-a93c-7ba3d045f543: issues processed=1979 output=1976

[GhZh Concatenate Issue]
- ghzh_issue bc2cb370-249c-4401-97b1-5471008a97ec issues=8540

[GhZh Update Issue History] ghah_batchid didn't increment
Input
- ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59 issues=17240
- ghzh_batchid 4e88a075-3fd1-4e57-9771-91d94ba90b42 value=17
- ghzh_issue bc2cb370-249c-4401-97b1-5471008a97ec issues=8540

Output
- ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59 issues=25780
- ghzh_batchid 4e88a075-3fd1-4e57-9771-91d94ba90b42 value=17

190205
Rerun.  [GhZh Update Issue History] includes concatenation of ghzh_issue1-3
GhZhToDomo.py
GhZhToDomo1: repos 0 to 10 > ghzh_issue1 6cd67468-4d9e-41cf-a135-832226a3f012: issues processed=3368 output=3368
GhZhToDomo2: repos 11 to 100 > ghzh_issue2 197b88a0-f8b3-4d97-af8d-27231b5f7f3c: issues processed=3215 output=3213
GhZhToDomo3: repos 101 to 300- > ghzh_issue3 a9f405b3-2958-4274-a93c-7ba3d045f543: issues processed=1980 output=1979
All ran successfully but output less issues than processed.

[GhZh Update Issue History] 
Input
    GhZhToDomo1
    GhZhToDomo2
    GhZhToDomo3
    ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59 issues=34320
    ghzh_batchid 4e88a075-3fd1-4e57-9771-91d94ba90b42 value=38
Output
    ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59 issues=
    ghzh_batchid 4e88a075-3fd1-4e57-9771-91d94ba90b42 value=

---

@BruceP99166 commented on [Sat Feb 02 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-460000297)

GhZhToDomo issues processed > issues output

---

@BruceP99166 commented on [Mon Feb 04 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-460021731)

190202 GhZh Update Issue History does not increment ghzh_batchid

190204 
Tried several times; seems to be working.  Value=19
Try running 3-part update.  Should increment to 22.

---

@BruceP99166 commented on [Wed Feb 06 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-461126142)

Running update with 3 repo subsets
GhZhToDomo.py
GhZhToDomo1: repos 0 to 10 > ghzh_issue1 6cd67468-4d9e-41cf-a135-832226a3f012: issues processed=3369 output=3368
    Failed to access ghzh_batchid.  Output less issues than processed.
GhZhToDomo2: repos 11 to 100 > ghzh_issue2 197b88a0-f8b3-4d97-af8d-27231b5f7f3c: issues processed=3221 output=3218 
    Failed to access ghzh_batchid.  Output less issues than processed.
GhZhToDomo3: repos 101 to 300- > ghzh_issue3 a9f405b3-2958-4274-a93c-7ba3d045f543: issues processed=1993 output=1992 
    Failed to access ghzh_batchid.  Output less issues than processed.

[GhZh Update Issue History]
Input
GhZhToDomo1
GhZhToDomo2
GhZhToDomo3
ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59 issues=42892
ghzh_batchid 1537db1c-8a09-4bd6-b3ee-c8f31f348824 value=54
Output
ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59 issues=51470
ghzh_batchid 9ff2e956-b982-475f-82db-17daaf7d6deb value=55
Update successful but new ghzh_batchid created.

---

@BruceP99166 commented on [Thu Feb 07 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-461532286)

GhZhToDomo creates new ghzh_batchid file instead of updating input file.  
From the dojo: https://dojo.domo.com/t5/Beast-Mode-ETL-Dataflow/Recursive-dataflow-not-storing-historical-data/m-p/33834
says that, for the same dataset to be used as both input and output, the dataset must first be created by the dataflow, then can be selected as input.

Run [GhZh Increment Batchid], created ghzh_batchid 7bac8447-cc8b-4704-8b24-b176fe3caccb value=56
Make ghzh_batchid 7bac8447-cc8b-4704-8b24-b176fe3caccb the input.
Rerun ghzh_batchid 7bac8447-cc8b-4704-8b24-b176fe3caccb value=57
Rerun ghzh_batchid 7bac8447-cc8b-4704-8b24-b176fe3caccb value=58

Run [GhZh Update Issue History] with input ghzh_batchid 9ff2e956-b982-475f-82db-17daaf7d6deb as input value=55.  ghzh_batchid update worked correctly.

---

@BruceP99166 commented on [Fri Feb 08 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-461965058)

*Output issue ds fails as input*
Created ghzh_issue_debug_190207 with dummy issue names and urls.  Verified that it fails as input to a dataflow.  Shared with Will Evans for debug.

---

@BruceP99166 commented on [Mon Feb 11 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-462388126)

Removing bad records from ghzh_issue_history
Run [GhZh Issue History Remove Bad Records] to select issues with batchId=55
Input 
ghzh_issue_history c88c9563-4be9-4acf-917e-e1f5c8b8fb59 rows=77247
Output
ghzh_issue_history 69a663c6-b666-4ccc-a2d7-a84d0a8dbbee rows=8578

Run [GhZh Update Issue History] to create ghzh_issue_history with issue rows from 2/8 appended
Input
ghzh_issue_history 69a663c6-b666-4ccc-a2d7-a84d0a8dbbee rows=8578
ghzh_batchid 9ff2e956-b982-475f-82db-17daaf7d6deb value=58
ghzh_issue1 6cd67468-4d9e-41cf-a135-832226a3f012 rows=3368
ghzh_issue2 197b88a0-f8b3-4d97-af8d-27231b5f7f3c rows=3229
ghzh_issue3 a9f405b3-2958-4274-a93c-7ba3d045f543 rows=2003
Output
ghzh_issue_history 6f8ff952-a7ce-4c86-ace6-403acf89fda1 rows=17200
   Should be rows=17178.  May have appended to existing 6f8f.
ghzh_batchid 63715feb-df49-4b8d-9014-48a5ccddb399 value=59

Defer until after including PRs in update.

---

@BruceP99166 commented on [Mon Feb 18 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-463781817)

**Output issue ds fails as input**
After adding PRs, issue2 ds has 16000 rows; fails as dataflow input.  Sent Will an email to tell him the issue has reappeared.  He replies that the ds ghzh_issue_debug_190208 I sent for his testing has 0 rows!?

Testing to see what ds size will work.
- ghzh_test_4000 rows is ok
- ghzh_test_12000 rows is running.   To rename from ghzh_test_ds_issue_length.

Email from Will Evans
```
When you create, load, and update a dataset in Domo, it keeps iterative versions of it as version control so that you can roll back and revert. As such, when you’re creating this dataset and then individually uploading rows, it creates a lot of overhead in the dataset, which we think is why the smaller datasets are working (smaller overhead), but the larger ones are not. We’re making some assumptions about the Domo backend here that we can’t directly validate, but everything points to this. When you try to load a dataset to run a dataflow, this operation will time out eventually causing the error you’re seeing. What we believe is happening is Domo is essentially trying to load 16,000 one row datasets, and the request eventually times out.

The fix for this is luckily pretty simple. In your code around line 725, you individually upload rows to Domo, we’d suggest refactoring this area of code to create a local CSV with 16,000 rows, and then upload it in one batch to Domo, like such:

        # Import Data from a file


        csv_file_path = './math.csv'

        datasets.data_import_from_file(dataset['id'], csv_file_path)

        domo.logger.info("Uploaded data from a file to DataSet {}".format(

        dataset['id']))

This should result in a dataset with much less overheard, and have the benefit of making all the subsequent data flows run much faster. I’d also guess that it should improve the speed of your python script as well. If you have any questions please let us know, but I think that Sam has finally cracked this one. Thanks
```

**authentication bug**
Authentication appears to be failing for the request to get the next page (line 107)

---

@BruceP99166 commented on [Mon Feb 18 2019](https://github.com/openstax/organizational-agility/issues/32#issuecomment-464823509)

**authentication**
Fixed

**Output issue ds fails as input**
Implemented csv fix.  
Trying with all repos.  Appears to have worked.

**Comment by BruceP99166** [comment]
GhZhToDomo.py appears to run successfully with no errors but does not update ghzh_issues ds.  Researching how to make pydomo errors visible.

**Comment by BruceP99166** [comment]
Update last night was successful.
However, when I ran the update today using yesterday's ghzh_issue as input, none of the issues are found in history, so the estimates are coming from Zenhub.  
Try update run with one repo to test the issue lookup in yesterday's issue ds.  First generated the issue ds and then used it as input.  Issue lookup worked successfully on second run,.
Perhaps the contents of ghzh_issue got removed prior to kicking off today's update.
